File: src/main/java/net/ostis/jesc/ScClient.java
package net.ostis.jesc;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.SneakyThrows;
import net.ostis.jesc.model.request.ScRequest;
import net.ostis.jesc.model.response.ScResponse;

import java.net.URI;

public class ScClient implements AutoCloseable {

    private ScMachineWebSocketClient wsc;

    private ObjectMapper objectMapper = new ObjectMapper();

    @SneakyThrows
    public ScClient(String host, int port) {
        var wsUrl = String.format("ws://%s:%d/ws_json", host, port);
        wsc = new ScMachineWebSocketClient(
                new URI(wsUrl)
        );
        wsc.connectBlocking();
    }

    @SneakyThrows
    public ScResponse sendRequest(ScRequest scRequest) {
        wsc.send(objectMapper.writeValueAsString(scRequest));
        return objectMapper.readValue(wsc.getMessage(), ScResponse.class);
    }

    @Override
    @SneakyThrows
    public void close() {
        wsc.close();
    }
}

File: lib/src/main/kotlin/ostis/jesc/client/ScClient.kt
package ostis.jesc.client

import ostis.jesc.client.model.event.ScEvent
import ostis.jesc.client.model.request.ScRequest
import ostis.jesc.client.model.request.payload.ScRequestPayload
import java.io.Closeable

interface ScClient: Closeable {
    fun <T> sendRequest(request: ScRequest<ScRequestPayload>, responseType: Class<T>): T
    fun addEventHandler(handler: (ScEvent) -> Unit)
    fun removeEventHandler(handler: (ScEvent) -> Unit)
}